### PR TITLE
[4.0] Allow Preview to load in new _blank instead of modal

### PR DESCRIPTION
**Pull request by PhilETaylor**
Pull Request for Issue #20241 #20240 .

### Summary of Changes

Allow Preview toolbar buttons to be used as a modal or a link to target=_blank

### Testing Instructions

```php
$toolbar = Toolbar::getInstance();
$toolbar->preview('http://myJoomla.com ','myJoomla.com', true); // NEW- target=_blank link button
$toolbar->preview('http://myJoomla.com ','myJoomla.com', false); // like its always been = modal
$toolbar->preview('http://myJoomla.com ','myJoomla.com'); // like its always been = modal
```

Also see https://github.com/joomla/joomla-cms/issues/20418 for follow up notes

### Documentation Changes Required

None

### CODE REVIEW

**Comment by Septdir** [comment]
#20241 - about com_content not only about opening preview in a new window
#20240 -  for j3.x Although I will not oppose its closure. If this is realized in j4.   

Need Testing Instructions

**Comment by Septdir** [comment]
I have tested this item :white_check_mark: successfully on 63edb82b9fb7a55889139591d63f99d27446ab0f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20417">issues.joomla.org/tracker/joomla-cms/20417</a>.</sub>

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on 63edb82b9fb7a55889139591d63f99d27446ab0f

@PhilETaylor resolve the branch conflict and will be ready
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20417">issues.joomla.org/tracker/joomla-cms/20417</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20417">issues.joomla.org/tracker/joomla-cms/20417</a>.</sub>

File: libraries/src/Toolbar/CoreButtonsTrait.php [hunk 1]
```diff
@@ -44,20 +44,35 @@ public function divider(string $text = ''): SeparatorButton
...
 	 */
-	public function preview(string $url, string $text = 'JGLOBAL_PREVIEW'): PopupButton
+	public function preview(string $url, string $text = 'JGLOBAL_PREVIEW', $newWindow = false)
 	{
-		return $this->popupButton('preview', $text)
-			->url($url)
-			->iframeWidth(640)
-			->iframeHeight(480)
-			->icon('icon-eye');
+		switch ($newWindow)
```
Comment (yvesh) [inline libraries/src/Toolbar/CoreButtonsTrait.php]: Does this really have to be a switch? With a boolean? :-)

**Review by yvesh — CHANGES_REQUESTED** [review]
Can we do a if here? Without else (just return) ;)

**Comment by PhilETaylor** [comment]
a bit nit picking especially as its RTC already... 

having two return's instead of one is just as bad as using a switch. 

I have pushed a new code flow - hope you like this one.

If it were my code I would have done FALSE === $newWindow instead of $newWindow === false, but I have had my code changed when I correctly do the former before :)

**Comment by Quy** [comment]
Please retest.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20417">issues.joomla.org/tracker/joomla-cms/20417</a>.</sub>

**Comment by Septdir** [comment]
I have tested this item :red_circle: unsuccessfully on 163b39647cd0fb98d133dd61d5cf9b2fb6f382cf<br />
```
			\JLoader::register('ContentHelperPreview', JPATH_ADMINISTRATOR . '/components/com_content/helpers/preview.php');
				$url = \ContentHelperPreview::url($this->item);
				$toolbar->preview($url, \JText::_('JGLOBAL_PREVIEW'))
					->bodyHeight(80)
					->modalWidth(90);

				$toolbar->preview($url,'New window', true); // NEW- target=_blank link button
				$toolbar->preview($url,'always modal', false); // like its always been = modal
				$toolbar->preview($url,'old'); // like its always been = modal
```

Add two buttons New window (test in com_content)
screenshot_2018-05-18 articles edit - joomla 4 - administration


<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20417">issues.joomla.org/tracker/joomla-cms/20417</a>.</sub>

**Comment by Septdir** [comment]
This code 

```
if ($newWindow === true)
		{
			$button = $this->linkButton('link', $text)
				->url($url)
				->attributes(['target' => '_blank'])
				->icon('icon-eye');
		}
		else
		{
			$button = $this->popupButton('preview', $text)
				->url($url)
				->iframeWidth(640)
				->iframeHeight(480)
				->icon('icon-eye');

		}
```

This code will work correctly (I checked)

**Comment by Septdir** [comment]
I have tested this item :white_check_mark: successfully on 710e5427bb4328d69b34afb6dc2e5e05b74c0e9d<br /><br />Now work correct<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20417">issues.joomla.org/tracker/joomla-cms/20417</a>.</sub>

**Comment by carlitorweb** [comment]
Good catch @Septdir

**Comment by PhilETaylor** [comment]
> Good catch @Septdir

Yeah sorry about that, I was "supervising" my kid and her friend at the swimming pool, and coding in the heat and humidity of the viewing area of the public swimming pool... I have a lot of test toolbar buttons and so missed the fact that the re-code flow duplicated the first one.

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on 710e5427bb4328d69b34afb6dc2e5e05b74c0e9d<br /><br />@PhilETaylor  happens to all ;)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20417">issues.joomla.org/tracker/joomla-cms/20417</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20417">issues.joomla.org/tracker/joomla-cms/20417</a>.</sub>

**Comment by yvesh** [comment]
I meant it like that (else is not necessary).. (Just as a note for future PRs)

```php
if ($newWindow === true)
{
	return $this->linkButton('link', $text)
		->url($url)
		->attributes(['target' => '_blank'])
		->icon('icon-eye');
}

return $this->popupButton('preview', $text)
	->url($url)
	->iframeWidth(640)
	->iframeHeight(480)
	->icon('icon-eye');
```

**Comment by Septdir** [comment]
@yvesh then we can write so

```
return $newWindow === true ? $this->linkButton('link', $text)->url($url)->attributes(['target' => '_blank'])->icon('icon-eye')
	: $this->popupButton('preview', $text)->url($url)->iframeWidth(640)->iframeHeight(480)->icon('icon-eye');
```

P.S personally I do not like very much when return gets a few lines. Or two identical. IMHO It's better to return a variable.

**Comment by PhilETaylor** [comment]
**This thread sums up why it it is so hard to contribute to this project.**

The problem with all your comments are that they are PERSONAL PREFERENCE and OPINION. 

**There was nothing wrong with the original proposed change. It met all the[ REQUIREMENTS of the Joomla Code Style](https://github.com/joomla/coding-standards/blob/master/manual/php.md), was readable, documents itself, and worked.**

MY PREFERENCE AND OPINIONS were put to one side, to comply with the [Joomla Code Style](https://github.com/joomla/coding-standards/blob/master/manual/php.md).

MY PREFERENCE AND OPINIONS would be NOT to have two return statements in a single function, that should be avoided at all cost. using if/?/: should also be avoid if it is going to create unreadable, long lines of code.  

Using an if/else or a switch both meet the [Joomla Code Style ](https://github.com/joomla/coding-standards/blob/master/manual/php.md) and whereas my PERSONAL PREFERENCE was a switch (as the code layout is nice, and it SELF DOCUMENTS the backward compatibility) I gave way to accept the if/else.

We are not at RTC.. Im not changing the code again.

**Comment by yvesh** [comment]
@PhilETaylor that's why i wrote "for the next PR" :-) 

And this is not on preference, else is considered bad style when it can be avoided. (Just google avoid else, return early.. thousands of articles on it).

And it is not meant as critics, it was just a well-intentioned suggestion to improve further..

**Comment by PhilETaylor** [comment]
I dont see why I should code to YOUR preferences for "the next PR" or any PR.

**The Joomla Project has its Code style guidelines - anything that meets them should be acceptable.**

If you dont like the The Joomla Project preferences. then you can propose changes to the guidelines in a PR in the github repo for them, however just because its YOUR preference **doesnt mean everything else is wrong.** 

Just because there are loads of articles in google, doesnt mean those are right, or even compatible with the Joomla guidelines, now, or in the future.

**Comment by alikon** [comment]
early return as suggested is not a Code Style preference,  is much more a documented "tecnique" to write more clean&manageable code "_Object Calisthenics_" ...

**Comment by PhilETaylor** [comment]
Actually its still personal preference. For as many that advocate return early, there are the same number of people that advocate against it. 

As there is nothing in the Joomla Code Guidelines against or for it, **a PR should not be criticised against your own ideals, personal opinions and code preferences.**  This discourages contribution and also descends a quick fix in to a personal attacks and massive thread of unneeded opinion.

**Comment by alikon** [comment]
it wasn't my intention to discourage contributions...
i'm sorry if you felt that way

**Comment by PhilETaylor** [comment]
If you feel strongly on these points then replace the TODO in this file https://github.com/joomla/coding-standards/blob/master/manual/basic-guidelines.md with a PR with "any words of wisdom about PHP best practice, maybe even references for design patterns"

**Comment by alikon** [comment]
challenge accepted  see https://github.com/joomla/coding-standards/pull/236

**Comment by alikon** [comment]
@PhilETaylor i know what you mean, and i've experienced that feelings quite sometimes, but if all of us give up ?...
at least i can tell that i've tried :smile: 
https://www.youtube.com/watch?v=tCZxGG_4ins

**Comment by infograf768** [comment]
I like if/else.much more readable.

**Comment by laoneo** [comment]
Thanks @PhilETaylor to make J4 great.

**Pull request merged**